server/bleep/src/cache.rs

```diff
@@ -0,0 +1,260 @@
+
+        Ok(())
+    }
+
+    pub(crate) async fn delete_for_repo(&self, reporef: &RepoRef) -> anyhow::Result<()> {
+        let mut tx = self.db.begin().await?;
+        self.delete_all(&mut tx, reporef).await?;
+        tx.commit().await?;
+
+        Ok(())
+    }
+
+    async fn delete_all(
+        &self,
+        tx: &mut sqlx::Transaction<'_, Sqlite>,
+        reporef: &RepoRef,
+    ) -> anyhow::Result<()> {
+        let repo_str = reporef.to_string();
+        sqlx::query! {
+            "DELETE FROM file_cache \
+                 WHERE repo_ref = ?",
+            repo_str
+        }
+        .execute(&mut *tx)
+        .await?;
+
+        Ok(())
+    }
+}
+
+pub struct ChunkCache<'a> {
+    qdrant: &'a QdrantClient,
+    cache: scc::HashMap<String, FreshValue<Payload>>,
+}
+
+impl<'a> ChunkCache<'a> {
+    pub async fn for_file(
+        qdrant: &'a QdrantClient,
+        repo_ref: &'a str,
+        repo_name: &'a str,
+        relative_path: &'a str,
+    ) -> anyhow::Result<ChunkCache<'a>> {
+        let response = qdrant
+            .search_points(&SearchPoints {
+                limit: 10000,
+                vector: vec![0f32; semantic::EMBEDDING_DIMENSION],
+                collection_name: semantic::COLLECTION_NAME.to_string(),
+                offset: None,
+                with_payload: Some(WithPayloadSelector {
+                    selector_options: Some(with_payload_selector::SelectorOptions::Enable(true)),
+                }),
+                filter: Some(Filter {
+                    must: [
+                        make_kv_keyword_filter("repo_ref", repo_ref),
+                        make_kv_keyword_filter("repo_name", repo_name),
+                        make_kv_keyword_filter("relative_path", relative_path),
+                    ]
+                    .into_iter()
+                    .map(Into::into)
+                    .collect(),
+                    ..Default::default()
+                }),
+                with_vectors: Some(WithVectorsSelector {
+                    selector_options: Some(with_vectors_selector::SelectorOptions::Enable(true)),
+                }),
+                ..Default::default()
+            })
+            .await?
+            .result
+            .into_iter()
+            .map(Payload::from_qdrant);
+
+        let cache = scc::HashMap::default();
+        for payload in response {
+            _ = cache.insert(
+                payload.id.as_ref().unwrap().clone(),
+                FreshValue::stale(payload),
+            );
+        }
+
+        Ok(Self { qdrant, cache })
+    }
+
+    pub fn update_or_embed(
+        &self,
+        data: &'a str,
+        embedder: impl FnOnce(&'a str) -> anyhow::Result<Embedding>,
+        mut payload: Payload,
+    ) -> anyhow::Result<()> {
+        let id = {
+            let mut bytes = [0; 16];
+            bytes.copy_from_slice(&blake3::hash(data.as_ref()).as_bytes()[16..32]);
+            Uuid::from_bytes(bytes).to_string()
+        };
```

Comment:
does this reduce a 32-byte hash to a 16-byte uuid? how resistant is this to collisions?

Reply:
Since `blake3` is a cryptographic hash, it should be pretty resistant to collisions. Reducing the 256bit output to 128 bit would still require more than 2^64 points in the db for a 50% chance of collision (according to the birthday paradox), so I don't think this is something we need to worry about too much.